src/MemoryTracing/MemoryTracingUtils.cpp

```diff
@@ -55,75 +49,106 @@ SystemMemoryUsage ParseMemInfo(std::string_view meminfo_content) {
+  system_memory_usage.set_available_kb(kMissingInfo);
+  system_memory_usage.set_buffers_kb(kMissingInfo);
+  system_memory_usage.set_cached_kb(kMissingInfo);
+  system_memory_usage.set_pgfault(kMissingInfo);
+  system_memory_usage.set_pgmajfault(kMissingInfo);
+
+  // Extract system-wide memory usage from the /proc/meminfo file.
   ErrorMessageOr<std::string> reading_result = orbit_base::ReadFileToString("/proc/meminfo");
   if (reading_result.has_error()) {
     ERROR("%s", reading_result.error().message());
     return reading_result.error();
   }
+  GetValuesFromMemInfo(reading_result.value(), &system_memory_usage);
 
-  SystemMemoryUsage parsing_result = ParseMemInfo(reading_result.value());
-  parsing_result.set_timestamp_ns(current_timestamp_ns);
-  return parsing_result;
-}
-
-int64_t GetVmRssFromProcessStatus(std::string_view status_content) {
-  if (status_content.empty()) return kMissingInfo;
-
-  std::vector<std::string> lines = absl::StrSplit(status_content, '\n');
-  if (lines.empty()) return kMissingInfo;
-
-  for (std::string_view line : lines) {
-    if (absl::StartsWith(line, "VmRSS")) {
-      std::vector<std::string> splits = absl::StrSplit(line, ' ', absl::SkipWhitespace{});
-      // According to the kernel code
-      // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/proc/task_mmu.c,
-      // the size unit of the VmRSS field in the /proc/<PID>/status file is fixed to "kB", which
-      // implies 1024 Bytes. We report in "KB" in ProcessSystemUsage.
-      int64_t rss_kb;
-      if (splits.size() == 3 && splits[2] == "kB" && absl::SimpleAtoi(splits[1], &rss_kb)) {
-        return rss_kb;
-      }
-      return kMissingInfo;
-    }
+  // Extract system-wide page fault statistics from the /proc/vmstat file.
+  reading_result = orbit_base::ReadFileToString("/proc/vmstat");
+  if (reading_result.has_error()) {
+    ERROR("%s", reading_result.error().message());
+    return reading_result.error();
   }
+  GetValuesFromVmStat(reading_result.value(), &system_memory_usage);
 
-  return kMissingInfo;
+  return system_memory_usage;
+}
+
+void GetValuesFromProcessStat(std::string_view stat_content,
+                              ProcessMemoryUsage* process_memory_usage) {
+  if (stat_content.empty()) return;
+
+  // According to the kernel code
+  // https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/proc/array.c,
+  // the /proc/<PID>/stat file records 52 process status information on a single line, in a fixed
+  // order. We are intersted in the following fields:
+  //   Field index | Name   | Format | Meaning
+  //    24         | rss    | %ld    | # of pages the process has in real memory
+  //    10         | minflt | %lu    | # of minor faults the process has made
+  //    12         | majflt | %lu    | # of major faults the process has made
+  std::vector<std::string> splits = absl::StrSplit(stat_content, ' ', absl::SkipWhitespace{});
+  if (splits.size() != 52) return;
```

Comment:
Why returning here? If we are sure we always have 52 fields, I would print an error msg and if not I would like to know to think in the fields that matter for us are still there in the same order.

Lastly, again, I would reset the process_memory_usage.

Reply:
We initialize the values in the memory proto to  `kMissingInfo` when we create it. When calling the Get method to update corresponding fields in the memory proto, we only update the proto when the format of the file content is correct.

If the format of the file content is incorrect, we return without updating the corresponding fields in the memory proto. According to the source code, we can make sure that this file content contains 52 fields. 

Only in very wired case the format of this linux virtual file could be wrong. The '/proc/<pid>/stat/` file provides only values and there is no parameter name. If some fields are missing in the file, even though the order might be correct but we still cannot find the value we would like to extract.

Reply:
I agree we should error out if the format is incorrect. Should we return an error from this function, though? Or at leaset log an error here so it's easier to see why this failed should it fail?

Reply:
I changed to return an error message which explains the reason of failing to extract information from the file content, for instance, empty file content, wrong format in which line, or fail to extract the value. Thanks for the suggestion.

Reply:
Sounds great, thanks!